### PR TITLE
Add Completion Window Options

### PR DESCRIPTION
**Pull request by CKolkey**
Hi! Previously I opened #846 to see about exposing some configuration options of the completion window to users. With some of the recent changes you made, this became very simple to implement. Let me know if this is mergeable, or if I need to add anything else.

### CODE REVIEW

File: doc/cmp.txt [hunk 1]
```diff
@@ -527,6 +527,16 @@ window.{completion,documentation}.zindex~
   `number`
   The completion window's zindex.
 
+                                       *cmp-config.window.completion.col_offset*
+window.completion.col_offset~
+  `number`
+  Offsets the completion window relative to the cursor.
```
Comment (gegoune) [inline doc/cmp.txt:533]: What if cursor is on column lesser than `col_offset`? Perhaps it should take a function and one co.uld calculate offset taking cursor position (could make function receive it as an argument to make it easier) into consideration

Comment (CKolkey) [reply doc/cmp.txt:533]: With a `col_offset` of `-10`, and the cursor at the first column of a buffer, it won't go over the edge of the window:
Screenshot 2022-04-14 at 10 33 00

However, if there's a split, it'll be offset by the amount specified:
Screenshot 2022-04-14 at 10 35 34


IMO this is what I would expect.

Comment (gegoune) [reply doc/cmp.txt:533]: Ah, ok then. Assumed window could go behind visible viewport. Thanks for clarifying. I am wondering whether there is a use case in which function (you could always accept either integer or function) would still be useful.

Comment (CKolkey) [reply doc/cmp.txt:533]: I suppose it's possible, though not for my use case :) I just want to have icons on the left and line the text up nicely

**Comment by par4m** [comment]
excatly what i needed, is it possible to add this for Documentation window (relative to Completion window, not cursor) as well as cmdline ?

**Comment by hrsh7th** [comment]
I think we should consider the configuration option.

I think the following API's is more flexible.

```lua
cmp.setup {
  window = {
    documentation = {
      on_open_before = function(...)
      end
    }
  }
}
```

I'm not happy with the naming and interface above, but I think it's better than the current one.

**Comment by CKolkey** [comment]
I'm not sure I understand what you mean. Since there is already a declarative configuration API (`window.completion.border`, `window.completion.winhighlight`, etc), I think it would make more sense to continue in the same style. 

That being said, it's your project, so I'll be happy to implement it how you like once I understand what that is :)

**Comment by hrsh7th** [comment]
I think we should re-design the way to customize window appearance in the future.
But this PR is enough maintenable and worth it.

File: lua/cmp/config/default.lua [hunk 2]
```diff
@@ -85,6 +85,8 @@ return function()
       completion = {
         border = { '', '', '', '', '', '', '', '' },
         winhighlight = 'Normal:Pmenu,FloatBorder:Pmenu,CursorLine:PmenuSel,Search:None',
+        col_offset = 0,
+        side_padding = 1,
       },
```
Comment (hrsh7th) [inline lua/cmp/config/default.lua:90]: I think we also add the `col_offset` and `side_padding` support for `config/window.lua`.

Comment (CKolkey) [reply lua/cmp/config/default.lua:90]: I've added the fields and the ability to overwrite them via the table passed in (`opts`). Is there any more to it than that?

**Comment by hrsh7th** [comment]
LGTM. Thank you~!

**Comment by fitrh** [comment]
Sorry to comment out here, I don't know if this is a problem or something to be expected, but did you @CKolkey realize that if we set `side_padding` to 0, we get an extra space at the end of the item?

20220622_221731_full

Doing some quick patch like this

```diff
diff --git a/lua/cmp/view/custom_entries_view.lua b/lua/cmp/view/custom_entries_view.lua
index d3e97c9..a627f7c 100644
--- a/lua/cmp/view/custom_entries_view.lua
+++ b/lua/cmp/view/custom_entries_view.lua
@@ -143,7 +143,7 @@ custom_entries_view.open = function(self, offset, entries)
   vim.api.nvim_buf_set_option(entries_buf, 'modified', false)
 
   local width = 0
-  width = width + 1
+  width = width + config.get().window.completion.side_padding
   width = width + self.column_width.abbr + (self.column_width.kind > 0 and 1 or 0)
   width = width + self.column_width.kind + (self.column_width.menu > 0 and 1 or 0)
   width = width + self.column_width.menu + 1
```

Fix that behavior

20220622_222041_full

But since it's just a quick patch and I don't dig deep enough into the rendering stuff, I don't think it's the right solution.

**Comment by CKolkey** [comment]
No, I didn't realize. I think you could likely use the `side_padding` when calculating window width, and that would solve it.

**Comment by max397574** [comment]
@fitrh where can we find the config for that amazing looking cmp?

**Pull request merged**